x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/AllowWriteStep.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.core.ilm;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateObserver;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.common.settings.Settings;
+
+import java.util.Objects;
+
+/**
+ * Make shrunken index writable, if requested, by removing {@link IndexMetadata.APIBlock#WRITE } block on the index.
+ */
+public class AllowWriteStep extends UpdateSettingsStep {
```

Comment:
I would like to start the discussion about why you chose to override the `UpdateSettingsStep` and not to use it directly in the `ShrinkAction`.

Reply:
Hmm, yeah I'm not sure making a new step made the most sense. My reasoning was that I wanted the steps of `ShinkAction` to always be the same set of steps. To do this, I pushed the branch on `allowWriteAfterShrink` down into the `AllowWriteStep`, allowing the `AllowWriteStep` to be added as the last step, whether or not `allowWriteAfterShrink` was true.

The alternative would be to use `UpdateSettingsStep` and to use `allowWriteAfterShrink` to conditonally build the set of steps with or without the `UpdateSettingsStep`. Looking around, I see that [DeleteAction](https://github.com/elastic/elasticsearch/blob/0075c1fb1e82c51c747729924ce67efef6daadeb/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/DeleteAction.java#L88) (and others) behave like this. Do you think this is the way to go?

Reply:
I went ahead moved the condition into `ShrinkAction` and replaced `AllowWriteStep` with `UpdateSettingsStep`. It adds a bit of complexity to ShrinkAction, but seems worthwhile since it doesn't involve a whole new step and a bunch of code.